### ISSUE TITLE
Warnings for a simple pytest

### ISSUE THREAD

**Issue by aparamon** [aio-libs/aiohttp#3695]
A simple pytest test case works but produces warnings.

## Steps to reproduce
* conftest.py:
````py
import pytest

def pytest_addoption(parser):
    parser.addoption('--host', type=str, default='localhost')
    parser.addoption('--port', type=int, default=80)
    parser.addoption('--base', type=str, default='/')
````
* test.py:
````py
import pytest
import aiohttp
import urllib

class AsyncTestSession(aiohttp.ClientSession):
    """
    Rewrites request netloc (host:port) and prepends base_path.
    """
    def __init__(self, netloc, base_path='', *args, **kwargs):
        # TODO: set ssl parameter
        super().__init__(*args, **kwargs)
        self.netloc = netloc
        self.base_path = base_path

    def _request(self, method, url, *args, **kwargs):
        scheme, netloc, path, params, query, fragment = urllib.parse.urlparse(url)
        if path:
            path = '/'.join((self.base_path, path.lstrip('/')))
        url = urllib.parse.urlunparse((scheme, self.netloc, path, params, query, fragment))
        return super()._request(method, url, *args, **kwargs)

@pytest.fixture
@pytest.mark.asyncio
async def async_client(request):
    host = request.config.getoption('--host')
    port = request.config.getoption('--port')
    base = request.config.getoption('--base')
    return AsyncTestSession('{}:{}'.format(host, port), base)

@pytest.mark.asyncio
async def test(async_client):
    async with async_client.get('http://server/search') as r:
        assert r.status == 200
````
* pytest test.py --host=google.com
````
============================= test session starts =============================
platform win32 -- Python 3.7.1, pytest-4.0.2, py-1.7.0, pluggy-0.8.0
rootdir: D:\, inifile:
plugins: remotedata-0.3.1, openfiles-0.3.1, doctestplus-0.2.0, asyncio-0.10.0, arraydiff-0.3
collected 1 item

test.py .Unclosed client session
client_session: <test.AsyncTestSession object at 0x000002013756AB38>
Unclosed connector
connections: ['[(<aiohttp.client_proto.ResponseHandler object at 0x0000020138AEEAC8>, 425722.937)]']
connector: <aiohttp.connector.TCPConnector object at 0x000002013756ACF8>
                                                                [100%]

============================== warnings summary ===============================
test.py:5
  D:\test.py:5: DeprecationWarning: Inheritance class AsyncTestSession from ClientSession is discouraged
    class AsyncTestSession(aiohttp.ClientSession):

-- Docs: https://docs.pytest.org/en/latest/warnings.html
==================== 1 passed, 1 warnings in 0.67 seconds =====================
````
Test passes but produces warnings. The warnings are unexpected; it is not clear how to modify the code to eliminate them.

## Your environment

* Windows 10
* Python 3.7.2
* aiohttp 3.5.2

**Comment by asvetlov** [comment]
The warning is correct: you don't call `AsyncTestSession.close()` method.
Moreover, inheritance from `aiohttp.ClientSession` is discouraged, the feature will be removed in aiohttp 4.0

**Comment by aparamon** [comment]
@asvetlov Thanks for your prompt comment!
I think I understand what the warnings say. However I completely miss how to modify the minimal code to eliminate them. Might there be a documentation that could be helpful for this and similar cases? Or maybe you could provide a hint on where to call `AsyncTestSession.close()` and what is the replacement for inheritance?

**Comment by aparamon** [comment]
N.B: It is desired to keep
````py
@pytest.mark.asyncio
async def test(async_client):
    async with async_client.get('http://server/search') as r:
        assert r.status == 200
````
unmodified.

**Comment by aparamon** [comment]
Ah, `Unclosed client session` warning appeared easy to fix:
````py
@pytest.fixture
@pytest.mark.asyncio
async def async_client(request):
    host = request.config.getoption('--host')
    port = request.config.getoption('--port')
    base = request.config.getoption('--base')
    async with AsyncTestSession('{}:{}'.format(host, port), base) as s:
        return s
````
However `DeprecationWarning: Inheritance class AsyncTestSession from ClientSession is discouraged` is still unclear. https://github.com/aio-libs/aiohttp/issues/2691 reveals that you decided to deprecate inheritance altogether because of one risky use-case. But as `aiohttp.ClientSession` seemingly doesn't support something like http://docs.python-requests.org/en/master/user/advanced/#transport-adapters, overriding `_request()` remains the least-hacky way to hook into request processing. Or I'm missing something?

**Comment by asvetlov** [comment]
transport-adapters concept is interesting, I love to have something in aiohttp later.
But you want to *rewrite* host part for tests only.
Please take a look on https://github.com/aio-libs/aiohttp/blob/master/aiohttp/test_utils.py#L226-L399
The class does almost the same with the only exception: it accepts pre-started tested application instead of providing `host:port`

**Comment by aparamon** [comment]
@asvetlov Thanks for your pointer!
However `TestClient` doesn't quite cut it: only it accepts test application, not a network address. `AsyncTestSession` allows to transparently (for test-cases code) switch between localhost Apache and cloud-hosted app; the test case code just accesses "http://server/" regardless.
Please do not turn `DeprecationWarning` into error until `AsyncTestSession` can be cleanly implemented with something like adapters.

**Comment by asvetlov** [comment]
I guess you can implement your own test client looking on built-in class example.

**Comment by aparamon** [comment]
@asvetlov Do you seriously mean the *recommended* way is to re-declare all `aiohttp.ClientSession` public API (methods `get()`, `put()`, ..., `close()`, `__enter__()`, `__exit__()`, `__aenter__()`, `__aexit__()`) manually??

**Comment by asvetlov** [comment]
Ideally, test suite *can* be refactored to satisfy your case.
It requires a pull request with code, tests, and documentation. 
If you want to solve your problem *right now* -- copy-paste `TestClient` and implement what you need.

**Comment by aparamon** [comment]
@asvetlov Thanks for the discussion!

I wish to attract your attention to the fact that straightforward implementation of https://github.com/aio-libs/aiohttp/issues/2691 proves harming. Inheritance may be abused frequently but there are useful applications. In my example, overriding a single method is simpler, more readable, and less fragile compared re-declaring the complete `aiohhtp.ClientSession` interface (it is dozens of lines of boring code!). Inheritance is a well-established Pythonic idiom, and that `aiohhtp.ClientSession` restricts users from sub-classing is unexpected.

Might it be possible that implementation of https://github.com/aio-libs/aiohttp/issues/2691 is re-considered?

**Comment by asvetlov** [comment]
You want to use the undocumented method.
aiohttp prevents this kind of wild usage.
You are asking me for relaxing the restriction back to allow usage of *private implementation details*. 
It saves code lines of your code (but you still can code what you want with current aiohttp without hacking).
Do I understand you correctly?
The root of the problem is not an inheritance from `ClientSession` but `ClientSession._request` overriding.

P.S.
Please, consider ClientSession as a *sealed* or *final* class, as `dict` or `socket.socket` for example. Build your code *on top* of it, not on hacking it.
Otherwise, there is no possibility to establish a *public API* concept at all.
P.P.S.
Extending the public API is a way to evolve the library, making implementation details public is not.

**Comment by aparamon** [comment]
> Please, consider ClientSession as a _sealed_ or _final_ class, as `dict` or `socket.socket` for example. Build your code _on top_ of it, not on hacking it.

But with `dict` I can do
````py
class MyDict(dict):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.foo = 'bar'

d = MyDict()
print(d.foo)
````
While `aiohttp` warns me on
````py
import asyncio
import aiohttp

class AsyncTestSession(aiohttp.ClientSession):
    def __init__(self, netloc, base_path='', *args, **kwargs):
        # TODO: set ssl parameter
        super().__init__(*args, **kwargs)
        self.netloc = netloc
        self.base_path = base_path

async def main():
    async with AsyncTestSession('baz') as s:
        print(s.netloc)

asyncio.run(main())
````
Note that I don't touch private methods here!

----

I agree that overriding `_request` is hackey. The alternative of copying the complete `aiohttp.ClientSession` public interface is much less attractive though. Adapters are very welcome; but please do not turn warning into error until they are implemented.

**Comment by asvetlov** [comment]
On my work we always use the following design:
```
class APIClient:
    def __init__(self, base_url: URL):
        self._client = aiohttp.ClientSession()
        self._base_url = base_url

    async def close(self):
        await self._client.close()

    def _make_url(self, path: str):
        return self._base_url / path

    async def api_call(self, user: str, arg1: str, arg2: str) -> str:
        url = self._make_url('users') / user
        url = url.with_query(arg1=arg1, arg2=arg2)
        async with self._client.get(url) as resp:
            # check for errors here
            return await resp.text()
```
It allows to pass arbitrary base URL for testing.
aiohttp session is not exposed outside of the client.
As a benefit, you can build complex logic for error handling, authorization, etc.

Maybe you need something like this?

**Comment by aparamon** [comment]
Thanks for sharing the code!

But to my taste it's too abridged/restrictive. For example, your `api_call` is only capable of `GET` requests. I prefer to keep my tests as close to "real" code as possible, to the extend that large test case fragments can be used for demo/benchmarking unmodified. To call `GET` API, I want to write `client.get()`, as usual. Tests are documentation, and the reader of test cases should see the familiar methods of API access. When special objects have to be used in tests, it ruins the confidence in the code of real applications.

I love `requests` because it allows me to do just that. Actually, `aiohttp.ClientSession` does too, only discourages me. I find it unexpected and wrong: I'm pretty confident my approach is solid and should be encouraged, not discouraged :-)

**Comment by asvetlov** [comment]
aiohttp strongly prefer aggregation over inheritance in the library API.
This design decision has a very strong reason.
Aggregation allows changing the library internals with keeping backward compatibility in *controlled order* if people never use a private API outside of aiohttp itself.
Inheritance opens too many implementaion details. 
Inheritance freeze in death the library basically: I cannot change anything in aiohttp because people *may* use it and their code *might* be broken by any minor update.
Say, I consider `ClientSession._request` as an implementation detail. I'm free to change its signature, add-remove arguments and even drop the method entirely which breaks your code.

You are free to disagree but sorry, this is one of aiohttp design principles (which is not unique to aiohttp only).
I provided you several hints on how to adapt your code to aiohttp design. A little more verbose but future compatible with aiohttp in the long term.

Closing the issue.
